lib/auth/saml.go

```diff
@@ -352,36 +412,62 @@ type samlAuthResponse struct {
 	attributeStatements map[string][]string
 }
 
-func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, error) {
+const (
+	DiagInfoError                         = "common.error"
+	DiagInfoResult                        = "common.result"
+	DiagInfoCreateUserParams              = "common.createUserParams"
+	DiagInfoSAMLAttributesToRoles         = "SAML.attributesToRoles"
+	DiagInfoSAMLAttributesToRolesWarnings = "SAML.attributesToRolesWarnings"
+	DiagInfoSAMLAssertionInfo             = "SAML.assertionInfo"
+	DiagInfoSAMLAttributeStatements       = "SAML.attributeStatements"
+	DiagInfoSAMLTraitsFromAssertions      = "SAML.traits"
+	DiagInfoSAMLConnectorTraitMapping     = "SAML.connector_traits"
+)
+
+func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, *services.SAMLAuthRequest, error) {
 	ctx := context.TODO()
-	requestID, err := parseSAMLInResponseTo(samlResponse)
+
+	requestID, err := ParseSAMLInResponseTo(samlResponse)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, nil, trace.Wrap(err)
 	}
-	request, err := a.Identity.GetSAMLAuthRequest(requestID)
-	if err != nil {
-		return nil, trace.Wrap(err)
+
+	traceErr := func(msg string, errInfo error) {
+		_ = a.Identity.TraceSAMLDiagnosticInfo(ctx, requestID, DiagInfoError, msg, errInfo.Error())
 	}
-	connector, err := a.Identity.GetSAMLConnector(ctx, request.ConnectorID, true)
+
+	request, err := a.Identity.GetSAMLAuthRequest(requestID)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		traceErr("Failed to get SAML Auth Request", err)
+		return nil, request, trace.Wrap(err)
```

Comment:
Generally it is bad practice to return a a valid value  from function that returns err != nil. 

Could you consider decoupling `request, err := a.Identity.GetSAMLAuthRequest(requestID)` from `validateSAMLResponse` flow  for example: 
```go
id, err := getReuqestIDFromSAMLResponse(samlResp)
request, err := a.Idenity.GetSAMLAuthRequest(id)
re, err := validateSAMLResponse(samlResponse, request) 
```

Reply:
In principle I agree, except we are already doing it here:

https://github.com/gravitational/teleport/blob/master/lib/auth/saml.go#L407

```go
	// Calculate (figure out name, roles, traits, session TTL) of user and
	// create the user in the backend.
	params, err := a.calculateSAMLUser(connector, *assertionInfo, request)
	if err != nil {
		return re, trace.Wrap(err)
	}
	user, err := a.createSAMLUser(params)
	if err != nil {
		return re, trace.Wrap(err)
	}
```

Any refactoring would have to change that too, and I'm not sure this is worth it.

Still, I'll give it some thought and experiment if there is a way to express this cleanly.

Reply:
I have refactored 'validateSAMLResponse' to make it only return an error or a value, but never both.

The cost is an additional parameter to the function. Personally, I find the new code easier to read, but let me know what you think.